MIT-licensed sparse() parent method and expert driver

Sacha0 opened the pull request:
Followup to #13001, #14631, and #14702. This pull request replaces the LGPL-licensed `sparse` parent method with an MIT-licensed version and introduces an expert driver `sparse!` underlying that method (see documentation for details on the expert driver).

In accord with discussion in #12605, #9928, #9906, and #6769, this PR's `sparse` method does not automatically purge numerical zeros. Hence this pull request comments two tests of numerical-zero purging. Rather than simply being commented, perhaps those tests should remain with `dropzeros!(sparse(...))` in place of `sparse(...)` and a reference to this PR?

Hats off to those who contributed to the existing implementation! Matching its performance in all tested cases was challenging. [Benchmarks results](https://gist.github.com/Sacha0/30296d6a55506d0a88a0#file-mitsparse_benchres-jl), [benchmark code](https://gist.github.com/Sacha0/30296d6a55506d0a88a0#file-mitsparse_benchcode-jl); the results should be comprehensible without reading the code. **tl;dr**: In these benchmarks, this PR's implementation ranges from matching the existing implementation's performance to reducing runtime by ~30%, and seems to allocate approximately `Vector{Ti}(m)` + `Vector{Ti}(n)` less storage. Notably the sparser the matrix, the better this PR's implementation's relative performance. Though solving #13400 requires a fundamentally different algorithm, #13400's example serves as an extreme illustration of the preceding point:

``` julia
using Benchmarks
@benchmark sparse([1;100000000], [1;1], [1;1], 100000000, 1, Base.AddFun())
@benchmark mitsparse([1;100000000], [1;1], [1;1], 100000000, 1, Base.AddFun())
```

(with `mitsparse` this PR's implementation) yields

``` julia
================ Benchmark Results ========================
     Time per evaluation: 1.94 s [1.90 s, 1.98 s]
Proportion of time in GC: 9.71% [9.01%, 10.40%]
        Memory allocated: 2.24 gb
   Number of allocations: 13 allocations
       Number of samples: 4
   Number of evaluations: 4
 Time spent benchmarking: 9.67 s

================ Benchmark Results ========================
     Time per evaluation: 735.76 ms [723.62 ms, 747.89 ms]
Proportion of time in GC: 6.50% [5.98%, 7.03%]
        Memory allocated: 762.94 mb
   Number of allocations: 11 allocations
       Number of samples: 12
   Number of evaluations: 12
 Time spent benchmarking: 9.73 s
```

When editing base/sparse/csparse.jl to remove the existing methods, I folded all blocks to avoid peaking at the LGPL code. So someone should check that the removal was graceful given it was blind but for method signatures.

Apologies for responding slowly elsewhere while working on this!

Sacha0 committed:
Replace the LGPL-licensed sparse() parent method with an MIT-licensed version. See #13001 and #14631.

hayd commented:
This is great!

Travis shows a whitespace failure.

    git rebase --whitespace=fix HEAD~1

Sacha0 commented:
> Travis shows a whitespace failure.

Derp! Fixed, thanks @hayd!

KristofferC commented:
Very impressive work!

nalimilan commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,218 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+    sparse(I, J, V, [m, n, combine])
+
+Create a sparse matrix `S` of dimensions `m` x `n` such that `S[I[k], J[k]] = V[k]`. The
+  `combine` function is used to combine duplicates. If `m` and `n` are not specified, they
```
I think the common style does not add indentation here.

Sacha0 replied:
Fixed, thanks!

nalimilan commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,218 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+
+For additional documentation and an expert driver, see `Base.SparseArrays.sparse!`.
+"""
+function sparse{Tv,Ti<:Integer}(
```
Also, arguments are usually put on the first line (same below).

Sacha0 replied:
Placing arguments all on the first line makes the `function sparse{...}(...` and ` function sparse!{...}(... ` lines exceed the suggested line-length limit. For `sparse` the line overrun isn't too bad, so I placed the arguments on the first line as suggested. But doing the same for `sparse!` yields an exceedingly long line that is somewhat difficulty to parse, so for now I left that line as is --- broken into multiple lines illustrating the arguments' logical grouping. What is the right thing to do here? Exceed the line-length limit grossly, or break from the arguments-on-first-line style? Thanks!

tkelman replied:
Or break the grouping, and just have the first one or two arguments on the first line.

nalimilan replied:
Yeah, I just meant that the first argument should be on the first line, but you can organize the others as you find the most appropriate.

Sacha0 replied:
How is this? Indenting the second through fourth argument lines to `sparse!`'s opening parenthesis's level requires several additional lines, so instead I indented to one level beyond the function body. Should I keep this indentation, or should the indentation be strictly to the opening parenthesis? Thanks!

tkelman replied:
That's less standard through the codebase. I prefer body plus one indent for line continuations of function arguments, but a lot of other people seem to like aligning to the paren.

Sacha0 replied:
Cheers, thanks! Will keep the body-plus-one indentation absent further input.

nalimilan commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,218 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+    coolen = length(I)
+    if length(J) != coolen || length(V) != coolen
+        throw(ArgumentError(string("the first three arguments' lengths must match, ",
+            "length(I) (=$(length(I))) == length(J) (= $(length(J))) == length(V) (= ",
```
Missing two more spaces.

Sacha0 replied:
Fixed (assuming I understood this correctly), thanks!

nalimilan commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,218 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+        cscnzval = Vector{Tv}()
+
+        sparse!(I, J, V, m, n, combine, klasttouch,
+            csrrowptr, csrcolval, csrnzval,
```
Indentation issue here too.

Sacha0 replied:
Fixed (assuming I understood this correctly), thanks!

nalimilan commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,218 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+    csrrowptr::Vector{Ti}, csrcolval::Vector{Ti}, csrnzval::Vector{Tv},
+    csccolptr::Vector{Ti} )
+    sparse!(I, J, V, m, n, combine, klasttouch,
+        csrrowptr, csrcolval, csrnzval,
```
And here, and below.

Sacha0 replied:
Fixed (assuming I understood this correctly), thanks!

KristofferC commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,218 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+    end
+    # This completes the unsorted-row, has-repeats CSR form's construction
+
+    # Sweep through the CSR form, simultaneously (1) caculating the CSC form's column
```
*calculating

StefanKarpinski commented:
This is excellent work. Let's review the technical aspects of this and not focus too much on cosmetic issues, which can be fixed after merging this.

nalimilan commented:
> This is excellent work. Let's review the technical aspects of this and not focus too much on cosmetic issues, which can be fixed after merging this.

Well, these are clearly minor, but they are probably easier to correct while fixing more significant aspects than after merging (when nobody will care enough to make another PR).

Great work, BTW, though I'm not in position of evaluating it!

StefanKarpinski commented:
I just don't want to nitpick a great PR to death with cosmetic stuff, but yes, it can be easier to fix before merging. @Sacha0, if it's easy for you to fix the cosmetic stuff, great. If not, then we can just merge and fix after.

ViralBShah commented:
Nicely done!

tkelman commented:
There are test failures that obviously need fixing, so "just merge and fix cosmetic stuff later" is a bit premature.

StefanKarpinski commented:
Yes, obviously the test failures need to get fixed before merging, but let's focus on those rather than a slew of indentation and formatting issues.

Sacha0 commented:
I greatly appreciate the detailed review / feedback, cosmetic included :). Constructive criticism is a gift; the more and sharper, the more I learn, so much thanks @nalimilan! I will do my best to address the comments.

CI is brilliant. I completely missed the arnoldi failure when testing locally! Thanks!

Sacha0 commented:
I can reproduce the failure locally, but curiously `Base.runtests("linalg/arnoldi")` succeeds where `Base.runtest("linalg")` fails. Somehow the former call does not force bounds checking whereas the latter does?

Sacha0 commented:
The `linalg/arnoldi` test failure should be fixed. I still wonder about the difference between `Base.runtests("linalg/arnoldi")` and `Base.runtests("linalg")` mentioned above. Thanks!

tkelman commented:
If you're using `Base.runtests`, then I don't think anything unusual happens with bounds checking. More likely, running different sets of tests in the same process results in different random numbers.

Sacha0 commented:
> If you're using Base.runtests, then I don't think anything unusual happens with bounds checking. More likely, running different sets of tests in the same process results in different random numbers.

The issue was an out-of-bounds access. Running with `--check-bounds=yes` enabled reproduction under all conditions (at the REPL, via `Base.runtests("linalg/arnoldi")`, and via `Base.runtests("linalg")`). Without `--check-bounds=yes`, the issue appeared only during `Base.runtests("linalg")`. So it seems bounds checking state must have been involved somehow? Moreover, there are no random numbers involved in the offending line? Thanks!

tkelman commented:
`Base.runtests("linalg")` runs multiple processes, so it may be that the check-bounds flag only gets propagated to child processes when running multiple tests.

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,215 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+              "$(length(V)))")))
+    end
+
+    if m == 0 || n == 0 || coolen == 0
```
this should throw a `BoundsError` if elements of `I` or `J` are outside of `m`-by-`n`

Sacha0 replied:
Fixed in lines 421 and 440, thanks!

tkelman replied:
No, if `m == 0 || n == 0` this should not return successfully if the input indices are out of bounds

Sacha0 replied:
Ah, I see, thanks! Fixing now.

Sacha0 replied:
Fixed (I think), thanks!

Sacha0 replied:
I suppose the first check (`!isempty(I)`) suffices given the earlier check `length(I) == length(J) == length(V)`. But perhaps this is more clear. Thoughts?

tkelman replied:
Sure, that works. I think the error message can be the same as the non-empty case though, `row values I[k] must satisfy 1 <= I[k] <= m` etc. Also always good to add more tests for this kind of corner case.

Sacha0 replied:
After a little thought I might advocate for the distinct, explicit error message in now: The `I[k]`-specific error message may be confusing where, for example, `m == 2`, `n == 0`, and `(I,J,V)` = `([2, 1, 1.0])`.

tkelman replied:
The current error message that starts "where ..., any entry is necessarily out-of-bounds," reads to me as too meandering for an error message. I'd do something like this
```jl
if coolen != 0
    if n == 0
        throw(BoundsError("column indices J[k] must satisfy 1 <= J[k] <= n"))
    elseif m == 0
        throw(BoundsError("row indices I[k] must satisfy 1 <= I[k] <= m"))
    end
end
```

Sacha0 replied:
Beautiful. Copied verbatim. Thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,215 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+Parent of and expert driver for `sparse`; see `sparse` for basic usage. This method
+allows the user to provide preallocated storage for `sparse`'s intermediate objects and
+result as described below. This capability enables more efficient successive construction
+of SparseMatrixCSCs from coordinate representations, and also enables extraction of an
```
``` `SparseMatrixCSC`s ```

Sacha0 replied:
Fixed, thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,215 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+This method consists of three gross steps: (1) Counting-sort the provided coordinate
+representation into an unsorted-row CSR form including repeated entries. (2) Sweep through
+the CSR form, simultaneously calculating the desired CSC form's column-pointer array,
+detecting repeated entries, and repacking the CSR from with repeated entries combined;
```
CSR form

Sacha0 replied:
Fixed, thanks!

mbauman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,215 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+    @inbounds for k in 1:coolen
+        Ik, Jk = I[k], J[k]
+        if 1 > Jk || n < Jk
+            throw(BoundsError(string("row values I[k] must satisfy 1 <= I[k] <= m")))
```
`column indices J[k]…`, no?  This also covers the TODO comment below, doesn't it?

Sacha0 replied:
Fixed, thanks!

tkelman replied:
Seems not? Both error messages still say "row values" where I agree with @mbauman that one should say "row indices" and the other should say "column indices"

Sacha0 replied:
Good catch --- only the `I` -> `J` part of the comment registered. Thanks! Fixing now.

Sacha0 replied:
Solid now? Thanks!

mbauman commented:
This is really great.  The unexported `sparse!` method is indeed really unsafe with the large `@inbounds` blocks, as you nicely document.  My hunch is that checking the lengths of the input vectors wouldn't be all that expensive compared to the actual work that the function does.  But if you've tried that and it's too costly, then it may be a necessary tradeoff.

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,219 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+of `SparseMatrixCSC`s from coordinate representations, and also enables extraction of an
+unsorted-column representation of the result's transpose at no additional cost.
+
+This method consists of three gross steps: (1) Counting-sort the provided coordinate
```
three major steps? "gross" seems a bit off

Sacha0 replied:
Major it is :). Thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,219 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+    @inbounds for k in 1:coolen
+        Ik, Jk = I[k], J[k]
+        if 1 > Jk || n < Jk
+            throw(BoundsError(string("row values J[k] must satisfy 1 <= J[k] <= m")))
```
`1 <= J[k] <= n`

Sacha0 replied:
Fixed, thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,219 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+        if 1 > Ik || m < Ik
+            throw(BoundsError(string("row indices I[k] must satisfy 1 <= I[k] <= m")))
+        end
+        csrrowptr[I[k]+1] += 1
```
advantageous at all to reuse `Ik` or is the compiler already catching this?

Sacha0 replied:
Good catch, fixed, thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,219 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+    length(cscrowval) < cscnnz && resize!(cscrowval, cscnnz)
+    length(cscnzval) < cscnnz && resize!(cscnzval, cscnnz)
+
+    # Finally counting-sort the row and nonzero values from the CSR from into cscrowval and
```
CSR form into

Sacha0 replied:
Fixed, thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,219 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+        i < m && (origcsrrowptrip1 = csrrowptr[i+2])
+    end
+
+    # Compute the CSC form's colptrs and store them shifted forward by one in cscrowptr
```
in csccolptr ?

Sacha0 replied:
One too many ptrs floating around. Fixed, thanks!

tkelman commented:
There are a couple extra calls to `string` that I don't think are needed where the error messages haven't been split across multiple lines, otherwise all looks great to me. A handful more tests of corner cases would always be nice but could be done separately.

Worth running say Convex.jl or some other package that we know uses `sparse` a lot as a more real-world benchmark case?

Sacha0 commented:
> There are a couple extra calls to string that I don't think are needed where the error messages haven't been split across multiple lines,

Good catch. Think I just fixed those. Thanks!

> Worth running say Convex.jl or some other package that we know uses sparse a lot as a more real-world benchmark case?

I would be curious. I am also curious about @KristofferC's benchmark in https://github.com/JuliaLang/julia/issues/12981#issuecomment-138132553 / #12998. And also for `transpose` with the new implementation.

ViralBShah commented:
I like the idea of testing some `sparse` heavy benchmarks. Also, JuMP is perhaps also a good candidate?

ViralBShah commented:
@jrevels Just a thought. Is it possible to use your benchmarking infrastructure to automatically run PkgEvaluator on nanosoldier? It would be handy in such cases.

tkelman commented:
> Is it possible to use your benchmarking infrastructure to run PkgEvaluator on nanosoldier?

I've been planning on borrowing some of this https://github.com/JuliaCI/BenchmarkTrackers.jl/blob/master/src/base_ci.jl to do that.

I was notified over email (and CI confirms) that `BoundsError` does not take a string. There's an existing case of this bug here, not covered by any existing tests: https://github.com/JuliaLang/julia/blob/493f913431b103bc37a1b5343bcec0f154bea795/base/sparse/umfpack.jl#L36

Sacha0 commented:
> I was notified over email (and CI confirms) that BoundsError does not take a string.

Likewise, fixed, thanks! :)

hayd commented:
Should the single argument constructor for BoundsError be removed?

https://github.com/JuliaLang/julia/blob/ec609ccf27e87b51d5c3f316637b349f5f340bc4/base/boot.jl#L221
https://github.com/JuliaLang/julia/blob/ec609ccf27e87b51d5c3f316637b349f5f340bc4/base/essentials.jl#L24

tkelman commented:
> Also, JuMP is perhaps also a good candidate?

For correctness yes. Not so much for performance, JuMP only calls `sparse` in two non-testing, non-printing places.

jrevels commented:
> Is it possible to use your benchmarking infrastructure to run PkgEvaluator on nanosoldier? It would be handy in such cases.

Interesting idea, this could be a neat follow-up project once the current CI tracker gets where I want it to be...if somebody wants to toy with this before I get to it, [GitHub.jl's CommentListener type](https://github.com/JuliaWeb/GitHub.jl#handling-webhook-events) might be of use.

On the subject of this PR, I'll put the benchmarks here on the list to add to BaseBenchmarks.jl (the current sparse tests - taken from `test/perf` - check indexing, not construction).

Sacha0 commented:
> A handful more tests of corner cases would always be nice but could be done separately.

Thanks for the detailed review @tkelman! As a token of my appreciation, please accept these humble edge case tests.

Sacha0 commented:
> My hunch is that checking the lengths of the input vectors wouldn't be all that expensive compared to the actual work that the function does. But if you've tried that and it's too costly, then it may be a necessary tradeoff.

@mbauman I suspect you are correct, at least for most use cases. I would be happy going either way. My philosophy with the expert driver was to get out of the user's way as much as possible, checks included in case the user was doing something unanticipated, e.g. manipulating tiny sparse matrices in a tight loop. I've added a line to the documentation suggesting testing with `--check-bounds=yes` to partially address this concern. Thoughts? Thanks!

mbauman commented:
Given that it's not exported, I'm not too worried here.  I think it definitely needs those checks if it ever gets exported, but as it stands it's probably fine.

mbauman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,221 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+`length(cscrowval) >= nnz(S)` and `length(cscnzval) >= nnz(S)`; hence, if `nnz(S)` is
+unknown at the outset, passing in empty vectors of the appropriate type (`Vector{Ti}()`
+and `Vector{Tv}()` respectively) suffices, or calling the `sparse!` method
+neglecting `cscrowval` and `cscnzval`.
```
These arrays are currently not optional.

Sacha0 replied:
Fixed, thanks!

mbauman commented on base/sparse/sparsematrix.jl:
```diff
@@ -318,7 +318,221 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+intermediate CSR forms and require `length(csrrowptr) >= m + 1`,
+`length(csrcolval) >= length(I)`, and `length(csrnzval >= length(I))`. Input
+array `klasttouch`, workspace for the second stage, requires `length(klasttouch) >= n`.
+Optional input arrays `csccolptr`, `cscrowval`, and `cscnzval` constitute storage for the
```
There's another optional here.

Sacha0 replied:
The CSC arrays are currently optional; see the method definitions immediately following the main `sparse!` definition. Thanks!

Sacha0 replied:
But there is a method argument style issue with the immediately following definitions. Fixing. Thanks!

tkelman commented on test/sparsedir/sparse.jl:
```diff
@@ -303,7 +310,8 @@ mfe22 = eye(Float64, 2)
 # issue #5169
-@test nnz(sparse([1,1],[1,2],[0.0,-0.0])) == 0
+# @test nnz(sparse([1,1],[1,2],[0.0,-0.0])) == 0
+# commented following change to sparse() in #14798, also see #12605, #9928, #9906, #6769
```
This PR seems to have fallen off the radar, I think the only thing we were missing was running some package tests and comparing performance on real use cases.

Also for these tests where you've commented them out, I think it would be preferable to change the tested-for value, maybe leaving a comment that points to these issues as being the reason for the value to be what it is after this change.

Sacha0 replied:
> This PR seems to have fallen off the radar, I think the only thing we were missing was running some package tests and comparing performance on real use cases.

Thanks for the bump! To clarify, was the request for package testing and benchmarking with e.g. Convex.jl directed at me? Apologies if so, I misinterpreted the request. Also if so, how should I go about package testing? Similarly, what benchmarking with Convex.jl would you like to see? Not being a Convex.jl user, I do not know what would constitute a meaningful benchmark.



> Also for these tests where you've commented them out, I think it would be preferable to change the tested-for value, maybe leaving a comment that points to these issues as being the reason for the value to be what it is after this change.

Good call --- I will rebase, touch those tests up, and update the PR. Thanks again!

KristofferC replied:
I can try some benchmarks with my Finite Element code.

Sacha0 replied:
> I can try some benchmarks with my Finite Element code.

I would be delighted to see those, particularly if your code can benefit from using `sparse!` in place of `sparse`. I should have a new version up shortly --- in the process of building and testing. Thanks and best!

tkelman replied:
No it shouldn't be required for you to run such benchmarks, and convex.jl in particular may have issues on nightly. More directed at other reviewers and people who want to see this merged, we can merge this without doing much benchmarking ahead of time but having some more testing of it in advance would be good to be better informed in case there are any situations where this might end up leading to regressions.

KristofferC replied:
In my code, master and this branch performed almost identically. I don't have time to test the sparse! driver now but it is great it is available!

Sacha0 replied:
> In my code, master and this branch performed almost identically.

Cheers, thanks!

Sacha0 commented:
Rebased, touched up tests, and added a note to `sparse!`'s documentation making it clear that output arrays `csccolptr`, `cscrowval`, and `cscnzval` can alias input arrays `I`, `J`, and `V` in space-constrained environments.

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -335,7 +335,222 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+`combine` function is used to combine duplicates. If `m` and `n` are not specified, they
+are set to `maximum(I)` and `maximum(J)` respectively. If the `combine` function is not
+supplied, duplicates are added by default. All elements of `I` must satisfy
+`1 <= I[k] <= m`, and all elements of `J` must satisfy `1 <= J[k] <= n`.
```
there have been updates to this docstring that you need to incorporate.

make sure the signature is consistent between here and the rst docs and run `make docs` to make sure you aren't undoing recent changes

Sacha0 replied:
> there have been updates to this docstring that you need to incorporate.
> 
> make sure the signature is consistent between here and the rst docs and run make docs to make sure you aren't undoing recent changes

First time playing in the `doc` sandbox, so apologies in advance for inevitable mistakes. I updated the signature for `sparse` in doc/stdlib/arrays.rst and issued `make julia-genstdlib`. No complaints about `sparse`, though complaints about other things:
```shell
UndefVarError(:build_sysimg)
WARNING: Mod Base build_sysimg
INFO: devdocs/sysimg.rst: no docs for build_sysimg(sysimg_path=default_sysimg_path, cpu_target="native", userimg_path=nothing; force=false) in Base

WARNING: Exported method missing doc for Base.StackTraces.StackTrace
WARNING: Exported method missing doc for StackFrame
WARNING: Exported method missing doc for Base.LibGit2.with
WARNING: Exported method missing doc for Base.Docs.doc
WARNING: Exported method missing doc for @threadcall

WARNING: Missing 5 exported doc strings
INFO: Missing 121 unexported doc strings
```
Guessing that constitutes success, there being no complaints about `sparse`? Subsequently issued `make docs`, and then `make -C doc html` and `make -C doc latex` for good measure; all seemed to complete fine. Should that do the trick? Will push after clarifying the `sparse!` docs. Thanks again!

tkelman replied:
check the diff on the rst and make sure you're incorporating rather than overwriting the recent updates to this docstring

Sacha0 replied:
Thanks for the pointer! Diff seems solid?

```diff
diff --git a/doc/stdlib/arrays.rst b/doc/stdlib/arrays.rst
index 5847201..ee0c291 100644
--- a/doc/stdlib/arrays.rst
+++ b/doc/stdlib/arrays.rst
@@ -855,11 +855,13 @@ Sparse Vectors and Matrices
 Sparse vectors and matrices largely support the same set of operations as their
 dense counterparts. The following functions are specific to sparse arrays.

-.. function:: sparse(I,J,V,[m,n,combine])
+.. function:: sparse(I, J, V,[ m, n, combine])

    .. Docstring generated from Julia source

-   Create a sparse matrix ``S`` of dimensions ``m x n`` such that ``S[I[k], J[k]] = V[k]``\ . The ``combine`` function is used to combine duplicates. If ``m`` and ``n`` are not specified, they are set to ``maximum(I)`` and ``maximum(J)`` respectively. If the ``combine`` function is not supplied, ``combine`` defaults to ``+`` unless the elements of ``V`` are Booleans in which case ``combine`` defaults to ``|``\ . All elements of ``I`` must satisfy ``1 <= I[k] <= m``\ , and all elements of ``J`` must satisfy ``1 <= J[k] <= n``\ .
+   Create a sparse matrix ``S`` of dimensions ``m x n`` such that ``S[I[k], J[k]] = V[k]``\ . The ``combine`` function is used to combine duplicates. If ``m`` and ``n`` are not specified, they are set to ``maximum(I)`` and ``maximum(J)`` respectively. If the ``combine`` function is not supplied, ``combine`` defaults to ``+`` unless the elements of ``V`` are Booleans in which case ``combine`` defaults to ``|``\ . All elements of ``I`` must satisfy ``1 <= I[k] <= m``\ , and all elements of ``J`` must satisfy ``1 <= J[k] <= n``\ .
+
+   For additional documentation and an expert driver, see ``Base.SparseArrays.sparse!``\ .
```

tkelman replied:
yep. looks like that isn't pushed yet?

Sacha0 replied:
Added the doc modification below and pushed. Thanks!

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -335,7 +335,222 @@ function sparse_IJ_sorted!{Ti<:Integer}(I::AbstractVector{Ti}, J::AbstractVector
+is wise.
+
+Output arrays `csccolptr`, `cscrowval`, and `cscnzval` can alias input arrays `I`, `J`,
+and `V` if space is limited.
```
what does "if space is limited" mean here? this isn't very clear

Sacha0 replied:
Would "if memory is limited" be clear? Suggestions? Thanks!

tkelman replied:
not really. is aliasing expected to happen, or something that users are allowed to do? the inputs and outputs are often going to be in very different orders...

Sacha0 replied:
Thanks, I think I see the opacity now. Is

> Where memory is limited, you may reuse the input arrays' storage (`I`, `J`, `V`) for the output arrays (`csccolptr`, `cscrowval`, `cscnzval`). For example, you may call `sparse!(I, J, V, csrrowptr, csrcolval, csrnzval, I, J, V)`.

better?

Tangentially, potential issue: The `SparseMatrixCSC` constructor does not enforce `length(S.colptr) == S.n + 1`. Meanwhile the idiom `S.colptr[end]` appears a few (six) times in base, potentially causing a subtle bug where `length(S.colptr) != S.n + 1`. Perhaps rewriting those appearances as `S.colptr[S.n + 1]` would be safer? If not, perhaps the `SparseMatrixCSC` constructor should enforce `length(S.colptr) == S.n + 1` alongside `m, n >= 0`? Thanks!

tkelman replied:
Or "to avoid additional allocation"? Or just say it's allowed without implying why the user might want to do it.

There are some use cases where wanting to construct a SparseMatrixCSC that is temporarily invalid in terms of the sizes and/or contents of the colptr, rowval, nzval fields is useful, so I'm not sure where such checks should be made.

Sacha0 replied:
> There are some use cases where wanting to construct a SparseMatrixCSC that is temporarily invalid in terms of the sizes and/or contents of the colptr, rowval, nzval fields is useful, so I'm not sure where such checks should be made.

Agreed. So perhaps replacing `S.colptr[end]` instances with `S.colptr[S.n + 1]` is wise? Improves safety without sacrificing flexibility?

tkelman replied:
as long as the performance difference is negligible

Sacha0 replied:
> as long as the performance difference is negligible

I will add this to the PRs-to-prepare list.

tkelman commented:
Ah, we should really document the change in behavior regarding not dropping zero values any more. And maybe it would be a good idea to add new copies of those changed tests but exercising `dropzeros!`.

Sacha0 commented:
> Ah, we should really document the change in behavior regarding not dropping zero values any more.

Sure, a line at the end of `sparse`'s documentation noting it retains numerical zeros as structural nonzeros? Or something else?

> And maybe it would be a good idea to add new copies of those changed tests but exercising dropzeros!

Shall add these tomorrow. Best!

tkelman commented:
Yeah, in sparse's documentation, and I think this is also worthy of mentioning in NEWS.md.

Sacha0 commented:
The Appveyor i686 discontent (failure on 7f845ffa81aa6f9b38827587521eae0ceb9bf436/test/arrayops.jl#L1060) seems unrelated?

Sacha0 commented:
> Yeah, in sparse's documentation, and I think this is also worthy of mentioning in NEWS.md.

Added a line tot t

Sacha0 commented:
Sorry, accidental close.

Iadded a line to `sparse`'s documentation and a mention in NEWS.md regarding the numerical zero retention change. I also added the `dropzeros!`-exercising test versions mentioned above. Best!

Sacha0 commented:
Sorry, I accidentally closed this and pushed a revision prior to noticing. GH will not allow me to reopen this pull request, and a brief search seems to indicate there is no recourse. How should I proceed?

Concerning the push, I added a line to `sparse`'s documentation and a mention in NEWS.md regarding the numerical zero retention change. I also added the `dropzeros!`-exercising test versions mentioned above. Thanks, and best!

StefanKarpinski commented:
Yeah, looks like it can't be reopened by an admin either so you'll have to make a new PR :-\

Sacha0 commented:
Opened anew (#15242). Apologies for the glitch!
